### PR TITLE
Feature/rwscmd fix

### PR DESCRIPTION
**Pull request by glow-mdsol**
@isparks @anewbigging @rsayer-mdsol FYI
Bit of wrangling with the rwscmd code
* added missing file and dependencies
* added Python3 support to rwscmd code
* added environment fix to get click tests to work
* moved tests into central testing location, cleared out unneeded files
* added travis-ci build task (see [https://travis-ci.org/mdsol/rwslib] for details)

### CODE REVIEW

File: rwslib/extras/rwscmd/odmutils.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+__author__ = 'anewbigging'
+
+from lxml import etree
+from enum import Enum
+
+# Constants
+ODM_NS = '{http://www.cdisc.org/ns/odm/v1.3}'
+MDSOL_NS = '{http://www.mdsol.com/ns/odm/metadata}'
+
+NSMAP = {None : "http://www.cdisc.org/ns/odm/v1.3", "mdsol": "http://www.mdsol.com/ns/odm/metadata"}
```
Comment (isparks) [inline rwslib/extras/rwscmd/odmutils.py]:
```
NSMAP = {None: ODM_NS, "mdsol" : MDSOL_NS}
```
?

File: rwslib/extras/rwscmd/odmutils.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    """Value prefix with ODM Namespace"""
+    return "{0}{1}".format(ODM_NS, value)
+
+def mdsol(value):
+    """Value prefix with mdsol Namespace"""
+    return "{0}{1}".format(MDSOL_NS, value)
+
+
+class E_ODM(Enum):
+    CLINICAL_DATA = odm('ClinicalData')
```
Comment (isparks) [inline rwslib/extras/rwscmd/odmutils.py]: Docstring:
```
"Defines ODM Elements"
```

?

File: rwslib/extras/rwscmd/odmutils.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    ITEM_DEF = odm('ItemDef')
+    RANGE_CHECK = odm('RangeCheck')
+    CODELIST_REF = odm('CodeListRef')
+    CODELIST = odm('CodeList')
+    CODELIST_ITEM = odm('CodeListItem')
+    ENUMERATED_ITEM = odm('EnumeratedItem')
+
+
+class A_ODM(Enum):
+    AUDIT_SUBCATEGORY_NAME = mdsol('AuditSubCategoryName')
```
Comment (isparks) [inline rwslib/extras/rwscmd/odmutils.py]: Docstring:

```Defines ODM Attributes```

?

File: rwslib/extras/rwscmd/rwscmd.py [hunk 4]
```diff
@@ -46,14 +47,14 @@ def get_data(ctx, study, environment, subject):
     """Call rwscmd_getdata custom dataset to retrieve currently enterable, empty fields"""
     studyoid = study + '(' + environment + ')'
     path = "datasets/" + GET_DATA_DATASET + "?StudyOID=" + studyoid + "&SubjectKey=" + subject + \
```
Comment (isparks) [inline rwslib/extras/rwscmd/rwscmd.py]: Urrgh....string concatenation.

/me falls to floor, twitching uncontrollably.

File: rwslib/extras/rwscmd/odmutils.py (hunk 1 above)
Comment (glow-mdsol) [inline rwslib/extras/rwscmd/odmutils.py]: Doesn't look to be used, dropping it out

File: rwslib/extras/rwscmd/odmutils.py (hunk 2 above)
Comment (glow-mdsol) [inline rwslib/extras/rwscmd/odmutils.py]: Done, thanks

File: rwslib/extras/rwscmd/odmutils.py (hunk 3 above)
Comment (glow-mdsol) [inline rwslib/extras/rwscmd/odmutils.py]: Done, thanks

File: rwslib/extras/rwscmd/rwscmd.py (hunk 4 above)
Comment (glow-mdsol) [inline rwslib/extras/rwscmd/rwscmd.py]: Switched to `"".format()`

File: rwslib/tests/test_rwscmd.py [hunk 5]
```diff
@@ -13,21 +15,18 @@ def setUp(self):
...
     @httpretty.activate
     def test_version(self):
-
         httpretty.register_uri(
             httpretty.GET, "https://innovate.mdsol.com/RaveWebServices/version",
             status=200,
             body='1.0.0')
 
-        result = self.runner.invoke(rwscmd.rws, ['https://innovate.mdsol.com','version'],
+        result = self.runner.invoke(rwscmd.rws, ['https://innovate.mdsol.com', 'version'],
```
Comment (isparks) [inline rwslib/tests/test_rwscmd.py]: This combination doesn't exist anywhere but do you want to give people ideas?

File: requirements.txt [hunk 6]
```diff
@@ -1,7 +1,9 @@
-lxml>=3.4.4
+lxml=3.6.2
```
Comment (isparks) [inline requirements.txt]: = or >= ?

File: rwslib/tests/test_rwscmd.py (hunk 5 above)
Comment (glow-mdsol) [inline rwslib/tests/test_rwscmd.py]: @anewbigging - one for you

File: rwslib/extras/rwscmd/rwscmd.py (hunk 4 above)
Comment (isparks) [inline rwslib/extras/rwscmd/rwscmd.py]: /me miraculously recovers.

**Comment by isparks** [comment]
Nice work, thanks Geoff !

File: rwslib/tests/test_rwscmd.py (hunk 5 above)
Comment (anewbigging) [inline rwslib/tests/test_rwscmd.py]: Sorry, not understanding the question.

Comment (anewbigging) [inline rwslib/tests/test_rwscmd.py]: Got it.   Yes I think would be good to change.

File: requirements.txt (hunk 6 above)
Comment (glow-mdsol) [inline requirements.txt]: There's a regression in the lxml that pulls in a realtime library which works on linux, but deep-sixes the build on OS X, will check if that's been resolved

Comment (glow-mdsol) [inline requirements.txt]: Ok, looks to have been fixed in 3.6.4, restriction removed

File: rwslib/tests/test_rwscmd.py (hunk 5 above)
Comment (glow-mdsol) [inline rwslib/tests/test_rwscmd.py]: Ok, I'm being slow - I assume we're talking about passing a username/password to an unauth'ed endpoint?

**Comment by isparks** [comment]
Are we still waiting for resolution on my last comment. Will that be coming from @anewbigging or will you be changing, @glow-mdsol  ?

File: rwslib/tests/test_rwscmd.py (hunk 5 above)
Comment (glow-mdsol) [inline rwslib/tests/test_rwscmd.py]: OK, I set it so the username/password default to the empty string, and that if nothing is passed an unauth RWSConnection is configured.  Please check that you are ok with the solution

**Comment by glow-mdsol** [comment]
I think we're good, please merge if you are happy.

**Comment by isparks** [comment]
Don't think I have write access to do the Merge. Someone else?

**Pull request merged**